### PR TITLE
Extend observe() to emit std::function objects

### PR DESCRIPTION
**Pull request by geektoni**
This addresses #5026. I still need to work around some details, but this is kind of what I had in mind to make it work. At the moment it doesn't pass the tests though. @karlnapf @gf712

### CODE REVIEW

File: src/shogun/base/SGObject.h [hunk 1]
```diff
@@ -1129,6 +1129,37 @@ class SGObject: public std::enable_shared_from_this<SGObject>
...
+	 * @param step time step
+	 * @param name name of this function
+	 * @param method the function we want to emit
+	 */
+	template <class T>
+	void observe(
+		const int64_t step, std::string_view name,
+		std::function<T()> method) const
+	{
+		if (get_num_subscriptions() == 0)
```
Comment (karlnapf) [inline src/shogun/base/SGObject.h]: I guess we could have more involved filters here right?
Like. "check if user requested to observe this name"?

File: src/shogun/lib/observers/ObservedValueTemplated.h [hunk 2]
```diff
@@ -21,6 +21,23 @@ namespace shogun
...
+		 * @param value the observed value
+		 */
+		ObservedValueTemplated(
+		    const int64_t step, const std::string_view name, const T value)
+		    : ObservedValue(step, name), m_observed_value(value)
+		{
+			this->watch_param(
+			    name, &m_observed_value,
+			    AnyParameterProperties(
+			        "Unamed observed value", ParameterProperties::READONLY));
```
Comment (karlnapf) [inline src/shogun/lib/observers/ObservedValueTemplated.h:37]: un-named? (also there is a type)

File: tests/unit/lib/observers/ParameterObserverTest_unittest.cc [hunk 3]
```diff
@@ -41,6 +41,10 @@ class MockEmitter : public SGObject
 		observe<int32_t>(1, "a", 1, p);
 		observe<int32_t>(1, "b", 1, p2);
 		observe<int32_t>(1, "None", 1, p3);
+
+		std::function<int32_t()> lambda_function = []() { return 10; };
```
Comment (karlnapf) [inline tests/unit/lib/observers/ParameterObserverTest_unittest.cc]: I think this method should have something in it that changes its state. So that you later on can assert that it was *not* called

File: tests/unit/lib/observers/ParameterObserverTest_unittest.cc [hunk 4]
```diff
@@ -138,6 +142,18 @@ TEST_F(ParameterObserverTest, filter_all_property)
...
+TEST_F(ParameterObserverTest, filter_method)
+{
+	std::vector<std::string> method_params = {"method"};
+	std::shared_ptr<ParameterObserver> observer(
+	    new ParameterObserverLogger(method_params));
+	emitter.subscribe(observer);
+	emitter.emit_value();
+
+	EXPECT_EQ(observer->get<int32_t>("num_observations"), 1);
+    EXPECT_EQ(observer->get_observation(0)->get_any().as<int32_t>(), 10);
```
Comment (karlnapf) [inline tests/unit/lib/observers/ParameterObserverTest_unittest.cc]: nice.
Another test should be to test that the method is not called as written above

File: src/shogun/base/SGObject.h (hunk 1 above)
Comment (geektoni) [reply src/shogun/base/SGObject.h]: mmh it won't be exactly trivial to do so. SGObject does not store directly the observers, just their "subscriptions". The filtering is done automatically by RxCpp (have a look at the `SGObject::subscribe`).

File: src/shogun/lib/observers/ObservedValueTemplated.h (hunk 2 above)
Comment (geektoni) [reply src/shogun/lib/observers/ObservedValueTemplated.h:37]: I'm not sure about this constructor. I may change it later on.

File: tests/unit/lib/observers/ParameterObserverTest_unittest.cc (hunk 3 above)
Comment (geektoni) [reply tests/unit/lib/observers/ParameterObserverTest_unittest.cc]: It makes sense. I will update the tests accordingly.

File: src/shogun/lib/observers/ParameterObserver.cpp [hunk 5]
```diff
@@ -82,3 +82,17 @@ index_t ParameterObserver::get_num_observations() const
 {
 	return utils::safe_convert<index_t>(m_observations.size());
 };
+
+std::shared_ptr<ObservedValue> ParameterObserver::execute_lambda(const TimedObservedValue &value)
+{
+    try
+    {
+        auto func =
+            value.first->get_any().as<std::function<std::shared_ptr<ObservedValue>()>>();
```
Comment (geektoni) [inline src/shogun/lib/observers/ParameterObserver.cpp]: @gf712 I have a question here. Let's say my `Any` object contains an `ObservedValueTemplated` pointer object. The class `ObservedValueTemplated` inherits from `ObservedValue`. Ideally, I should be able to cast my `Any` to an `ObservedValue` pointer object right (e.g., `any.as<std::shared_ptr<ObservedValue>>()`)?

I thought it would have been possible to do so thanks to the inheritance, but it does not seem to work :/ is this an intended behaviour (or am I missing something)?

Comment (gf712) [reply src/shogun/lib/observers/ParameterObserver.cpp]: Yea, it doesn't work off the bat. You need to register that inheritance using
https://github.com/shogun-toolbox/shogun/blob/7014676a95ec27b51473f037c4c137083075d61e/src/shogun/lib/any.h#L1259-L1275

Comment (gf712) [reply src/shogun/lib/observers/ParameterObserver.cpp]: This is basically a vtable, but it is only populated with SGObject derived classes to SGObject and some primitive types + std::string. I guess you could always register it there? @vigsterkr

File: src/shogun/base/SGObject.h (hunk 1 above)
Comment (karlnapf) [reply src/shogun/base/SGObject.h]: I see....then we are kind of back to square one right?
At least the code is now here rather than in the rf model ....

Comment (geektoni) [reply src/shogun/base/SGObject.h]: Well, the main difference is that we can at least limit the execution of expensive functions now. 

For instance, in the BaggingMachine case, the `compute_oob()` method was called before emitting all observations. Now, the `compute_oob()` is called only if an observable wants it and it is executed by the observable rather than the model.

Comment (karlnapf) [reply src/shogun/base/SGObject.h]: ah maybe I got that wrong then. Cool! :)

File: tests/unit/lib/observers/ParameterObserverTest_unittest.cc [hunk 6]
```diff
@@ -41,7 +46,18 @@ class MockEmitter : public SGObject
 		observe<int32_t>(1, "a", 1, p);
 		observe<int32_t>(1, "b", 1, p2);
 		observe<int32_t>(1, "None", 1, p3);
+
+		std::function<int32_t()> lambda_function = [this]() {
+			this->member_variable++;
+			return 20;
+		};
+
+		observe<std::function<int32_t()>>(
```
Comment (geektoni) [inline tests/unit/lib/observers/ParameterObserverTest_unittest.cc:55]: Apparently, `Any` is smarter than I thought :) with the current implementation we can emit `std::function` directly without changing anything in the framework. We can also execute them in a fairly simple way by just calling `get<lambda return type>("name of the lambda")`. See the unit tests for an example.

Next question. Some of these operations could be quite costly and if the user calls them multiple times then they will be executed from scratch every time. Shall we somehow cache the result of these executions, so that the user will just need to call the lambda once and then it will retrieve only the pre-computed result?

This caching will happen only for the stored observations though.

Comment (karlnapf) [reply tests/unit/lib/observers/ParameterObserverTest_unittest.cc:55]: could you make an example of such an operation?

Comment (geektoni) [reply tests/unit/lib/observers/ParameterObserverTest_unittest.cc:55]: > could you make an example of such an operation?

mmh I don't have any real example now. I was thinking about possible use cases/corner cases that we may face.

Comment (gf712) [reply tests/unit/lib/observers/ParameterObserverTest_unittest.cc:55]: @geektoni I am modifying things a bit in #5037, so we use visitor patterns instead of `any_cast`. The only things you will have to keep in mind is that the visitor will only call a function when you get it if you have the `ParameterProperties::FUNCTION`, which is added in that PR. About the caching, you could use a `static` variable inside the lambda.

Comment (gf712) [reply tests/unit/lib/observers/ParameterObserverTest_unittest.cc:55]: actually, maybe don't use static, that might cause massive issues

Comment (gf712) [reply tests/unit/lib/observers/ParameterObserverTest_unittest.cc:55]: If you want to cache actually, use `std::optional`. Put the result there. If it is set it returns the value, otherwise computes it.
